### PR TITLE
Step 948 input issue

### PR DESCRIPTION
**Pull request by godrei**
### Checklist

- [x] I've read and accepted the [Contribution Guidelines](https://github.com/bitrise-steplib/.github/blob/main/CONTRIBUTING.md)
- Requires _PATCH_ [version update](https://semver.org/)

### Context

This PR adds missing project_path step input validation.

Resolves: https://github.com/bitrise-steplib/steps-xcode-archive/issues/193

### Changes

Break Step into:
- step.ProcessInputs
- step.InstallDeps
- step.Run
- step.ExportOutput
functions.

Test and implement step.ProcessInputs's missing project_path extension validation.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -326,32 +375,118 @@ func main() {
...
+		} else if !installed {
+			log.Warnf(`xcpretty is not installed`)
+			fmt.Println()
+			log.Printf("Installing xcpretty")
+
+			cmds, err := xcpretty.Install()
+			if err != nil {
+				log.Warnf("Failed to create xcpretty install command: %s", err)
+				log.Warnf("Switching to xcodebuild for output tool")
+				// outputTool = "xcodebuild"
```
Comment (lpusok) [inline main.go:415]: The outputtool is not changed back at the moment.

File: main.go [hunk 2]
```diff
@@ -503,59 +670,67 @@ The log file is stored in $BITRISE_DEPLOY_DIR, and its full path is available in
...
+		archiveCodeSignIsXcodeManaged := profileutil.IsXcodeManaged(mainApplication.ProvisioningProfile.Name)
+
 		generator := NewExportOptionsGenerator(xcodeProj, scheme, configuration)
-		exportOptions, err := generator.GenerateApplicationExportOptions(exportMethod, cfg.ICloudContainerEnvironment, cfg.TeamID,
-			cfg.UploadBitcode, cfg.CompileBitcode, archiveCodeSignIsXcodeManaged, xcodeMajorVersion)
+		exportOptions, err := generator.GenerateApplicationExportOptions(exportMethod, opts.ICloudContainerEnvironment, opts.TeamID,
+			opts.UploadBitcode, opts.CompileBitcode, archiveCodeSignIsXcodeManaged, int64(opts.XcodeMajorVersion))
 		if err != nil {
-			fail(err.Error())
+			return fmt.Errorf(err.Error())
```
Comment (lpusok) [inline main.go:705]: return err ?

File: main.go [hunk 3]
```diff
@@ -77,9 +78,22 @@ type configs struct {
 	CacheLevel string `env:"cache_level,opt[none,swift_packages]"`
 }
 
-func fail(format string, v ...interface{}) {
-	log.Errorf(format, v...)
-	os.Exit(1)
+// Config ...
+type Config struct {
+	Inputs
```
Comment (mateherber) [inline main.go:83]: wow I never seen such syntax. What does this mean?
1) Config will have ExportMethod, UploadBitcode, etc. fields just like the Inputs (quasi-include)
2) Config will have an Input named Input typed field

?

File: main.go [hunk 4]
```diff
@@ -666,76 +950,163 @@ is available in the $BITRISE_IDEDISTRIBUTION_LOGS_PATH environment variable`)
...
+	}
+
+	installDepsOpts := InstallDepsOpts{
+		InstallXcpretty: config.OutputTool == "xcpretty",
+	}
+	if err := step.InstallDeps(installDepsOpts); err != nil {
+		return err
+	}
+
+	runOpts := RunOpts{
```
Comment (lpusok) [inline main.go:1051]: NOTE: I find it counter-intiutive that step.Run has no "outputs", only by side effects of placing files on predefined paths.

File: main.go [hunk 5]
```diff
@@ -139,98 +153,114 @@ func exportDSYMs(dsymDir string, dsyms []string) error {
 	return nil
 }
 
-func main() {
-	var cfg configs
-	if err := stepconf.Parse(&cfg); err != nil {
-		fail("Issue with input: %s", err)
+// XcodeVersionProvider ...
+type XcodeVersionProvider interface {
```
Comment (lpusok) [inline main.go:157]: Is there a specific reason to use exported functions/structs?

**Review by mateherber — APPROVED** [review]

File: main.go [hunk 6]
```diff
@@ -139,98 +153,114 @@ func exportDSYMs(dsymDir string, dsyms []string) error {
...
+	GetXcodeVersion() (models.XcodebuildVersionModel, error)
+}
+
+// XcodebuildXcodeVersionProvider ...
+type XcodebuildXcodeVersionProvider struct {
```
[the remark this answers is not in the document]
Comment (lpusok) [reply main.go:162]: xcodebuild is a command-line tool

[the remark this answers is not in the document]
Comment (lpusok) [reply main.go:162]: `DefaultXcodeVersionProvider` maybe is more readable

File: main.go (hunk 3 above)
Comment (godrei) [reply main.go:83]: https://golang.org/doc/effective_go#embedding

With this type of embedding, Config will borrow the field and methods of Input.

File: main.go (hunk 5 above)
Comment (godrei) [reply main.go:145]: No

File: main.go [hunk 7]
```diff
@@ -77,9 +78,22 @@ type configs struct {
...
+	IDEDistributionLogsZipPath string
+	XcodebuildLogPath          string
+	ExportOptionsPath          string
+	IPAPath                    string
+	ArchiveZipPath             string
+	DSYMZipPath                string
+	AppPath                    string
+	IPAExportDir               string
+
+	XcodeMajorVersion int
```
[the remark this answers is not in the document]
Comment (godrei) [reply main.go:84]: What I wanted to achieve is logically grouping the inputs.

File: main.go [hunk 8]
```diff
@@ -139,98 +153,114 @@ func exportDSYMs(dsymDir string, dsyms []string) error {
...
+type StepInputParser interface {
+	Parse(conf interface{}) error
+}
+
+// EnvStepInputParser ...
+type EnvStepInputParser struct {
+}
+
+// NewEnvStepInputParser ...
+func NewEnvStepInputParser() EnvStepInputParser {
```
[the remark this answers is not in the document]
Comment (godrei) [reply main.go]: unified the code to return descendants

**Review by mateherber — APPROVED** [review]

**Review by lpusok — APPROVED** [review]

**Pull request merged**